libcxx/test/tools/clang_tidy_checks/robust_against_operator_ampersand.cpp

```diff
@@ -0,0 +1,47 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang-tidy/ClangTidyCheck.h"
+#include "clang-tidy/ClangTidyModuleRegistry.h"
+#include "clang/Tooling/FixIt.h"
+
+#include "robust_against_operator_ampersand.hpp"
+
+// This clang-tidy check ensures that we don't use operator& on dependant
+// types. If the type is user supplied it may call the type's operator&.
+// Instead use std::addressof.
```

Comment:
This should go into the coding guidelines instead.

Reply:
I disagree. The code should documented what the intention is.

As for the coding guidelines; it already has this information [documented](https://libcxx.llvm.org/CodingGuidelines.html#don-t-use-argument-dependent-lookup-unless-required-by-the-standard).

(Note testing for `operator,` in the same part of the documentation is on my todo list.)

Reply:
I agree with @mordante here, this is documented explicitly in the coding guidelines already:

> Function overloading also applies to operators. Using &user_object may call a user-defined operator&. Use std::addressof instead. Similarly, to avoid invoking a user-defined operator,, make sure to cast the result to void when using the , or avoid it in the first place. For example:

IMO this comment doesn't provide absolutely necessary information, but it also doesn't hurt. I'd keep it.

Reply:
The guidelines should still document that this check exists. IMO this comment isn't useful, since it will get out of sync with the guidelines. I'd be much happier with a reference to the guidelines than documenting the same thing in different places.

Reply:
I would suggest the following, then:

```suggestion
// This clang-tidy check ensures that we don't use operator& on dependant
// types, as documented in our coding guidelines.
```

I am mostly neutral on this issue. It's true that we're duplicating a bit of information, but we're really not duplicating a lot of it. I'd let @mordante decide here to avoid blocking this whole patch on such a minor point.

Reply:
I don't think it's useful still, but I wouldn't block it on this. However, I would block on not mentioning the check in the documentation as we do with all other checks.